### PR TITLE
Make network-libp2p's Service generic over the message

### PR DESCRIPTION
**Pull request by tomaka**
~~Based on top of #1634, only the last commit is part of this PR.~~
First step of #1692.

We now encode and decode messages in the tokio task dedicated to each node.

### CODE REVIEW

File: core/network-libp2p/tests/test.rs [hunk 1]
```diff
@@ -16,12 +16,13 @@
...
 use std::{io, iter};
-use substrate_network_libp2p::{ServiceEvent, multiaddr};
+use substrate_network_libp2p::{CustomMessage, ServiceEvent, multiaddr};
 
 /// Builds two services. The second one and further have the first one as its bootstrap node.
 /// This is to be used only for testing, and a panic will happen if something goes wrong.
-fn build_nodes(num: usize) -> Vec<substrate_network_libp2p::Service> {
-	let mut result: Vec<substrate_network_libp2p::Service> = Vec::with_capacity(num);
+fn build_nodes<TMsg>(num: usize) -> Vec<substrate_network_libp2p::Service<TMsg>>
+where TMsg: CustomMessage + Send + 'static {
```
Comment (bkchr) [inline core/network-libp2p/tests/test.rs]:
```suggestion
	where TMsg: CustomMessage + Send + 'static
{
```

File: core/network/src/message.rs [hunk 2]
```diff
@@ -194,6 +197,17 @@ pub mod generic {
 		ChainSpecific(Vec<u8>),
 	}
 
+	impl<Header, Hash, Number, Extrinsic> CustomMessage for Message<Header, Hash, Number, Extrinsic>
+	where Self: Decode + Encode {
```
Comment (bkchr) [inline core/network/src/message.rs]:
```suggestion
		where Self: Decode + Encode
	{
```

File: core/network/src/message.rs [hunk 3]
```diff
@@ -194,6 +197,17 @@ pub mod generic {
...
+		fn into_bytes(self) -> Vec<u8> {
+			self.encode()
+		}
+
+		fn from_bytes(bytes: &[u8]) -> Result<Self, ()> {
+			Decode::decode(&mut Cursor::new(bytes)).ok_or(())
```
Comment (bkchr) [inline core/network/src/message.rs]:
```suggestion
			Decode::decode(&mut &bytes[..]).ok_or(())
```

File: core/network/src/service.rs [hunk 4]
```diff
@@ -537,28 +534,15 @@ fn run_thread<B: BlockT + 'static, S: NetworkSpecialization<B>>(
...
 				debug!(target: "sync", "{} clogging messages:", messages.len());
-				for msg_bytes in messages.iter().take(5) {
-					if let Some(msg) = Decode::decode(&mut (&msg_bytes as &[u8])) {
-						debug!(target: "sync", "{:?}", msg);
-						let _ = protocol_sender.send(ProtocolMsg::PeerClogged(node_index, Some(msg)));
-					} else {
-						debug!(target: "sync", "{:?}", msg_bytes);
-						let _ = protocol_sender.send(ProtocolMsg::PeerClogged(node_index, None));
-					}
+				for msg in messages.into_iter().take(5) {
```
Comment (bkchr) [inline core/network/src/service.rs]: Why do we throw all other messages away? Is that a wanted behavior?

Comment (arkpar) [reply core/network/src/service.rs]: It should really send it once per peer I think.

Comment (arkpar) [reply core/network/src/service.rs]: Also there's too much logging here. Line below logs the message and the protocol handler logs it again.

**Comment by gavofyork** [comment]
@tomaka waiting for @arkpar 's grumble to be sorted

File: core/network/src/service.rs (hunk 4 above)
Comment (tomaka) [reply core/network/src/service.rs]: I think it's only logged here. I don't see anywhere else where the messages are being printed.

Comment (bkchr) [reply core/network/src/service.rs]: @tomaka Could you also answer my question at the top?^^

**Comment by gavofyork** [comment]
Also waiting for @bkchr 's question to be answered :)

File: core/network/src/service.rs (hunk 4 above)
Comment (tomaka) [reply core/network/src/service.rs]: The `Clogged` message happens when the queue of messages is full. When that happens, we clone the content of the queue and send it back to the other API layer for "analysis". The code here is that "analysis", and just prints the first 5 messages in the logs.

After #1517, we will instead open one individual substream per message and this "clogged" message will therefore disappear.

**Pull request merged**